### PR TITLE
Case references

### PR DESCRIPTION
**Pull request by aaronbrowder**
This is a new feature which allows you to reference a previous JC case when creating a new case, for example if a sentence was broken. This makes it easier to retrieve information about previous cases and sentences when you need it. This feature really shines when you have a chain of broken sentences, as it enables you to see the full history of the issue without having to look anything up. Additionally, it allows you to decide what happens to the original sentence when a new sentence is created.

Hopefully this feature is pretty self-explanatory and you can learn how it works by playing with it. Since it affects the existing JC system, you may want to test that everything still works. Customers who do not enable the feature on the settings page should not experience any difference.

### CODE REVIEW

**Comment by aaronbrowder** [comment]
No problem! This is the first of several features I want to add. I'll have more coming soon, in both the JC and Attendance systems.

File: app/models/Case.java [hunk 1]
```diff
@@ -193,4 +217,147 @@ public String getRedactedFindings(Person keep_this_persons_name) {
...
+            return findings;
+        }
+        ArrayList<Charge> relevant_charges = new ArrayList<Charge>();
+        for (Charge charge : charges) {
+            charge.buildChargeReferenceChain(relevant_charges);
+        }
+        return generateCompositeFindings(relevant_charges, null);
+    }
+
+    private String generateCompositeFindings(List<Charge> relevant_charges, List<Case> used_cases) {
```
Comment (schmave) [inline app/models/Case.java]: This is a very cool function. I love the grouping by charge and resolution plan. Genius!

File: app/controllers/Application.java [hunk 2]
```diff
@@ -882,6 +891,30 @@ public static String jsonRules(String term) {
         return Json.stringify(Json.toJson(result));
     }
 
+    public static String jsonCases(String term) {
+        term = term.toLowerCase();
+
+        List<Case> cases = Case.find.where()
+            .eq("meeting.organization", Organization.getByHost())
+            .orderBy("case_number ASC").findList();
```
[the remark this answers is not in the document]
Comment (aaronbrowder) [reply app/controllers/Application.java]: It's weird that the names are blank. It's possible that whatever is causing that is also causing the other issues. That's just pulling from `charge.person.display_name`. Is `display_name` populated for your people?

File: app/views/view_settings.scala.html [hunk 3]
```diff
@@ -130,6 +143,39 @@ <h4>Other @OrgConfig.get().str_jc_name_short options</h4>
...
+        a new @OrgConfig.get().str_res_plan is created.
+    </p>
+    <p>
+        <label>
+            <input type="checkbox" name="enable_case_references" @if(org.enable_case_references){ checked }>
+            Enable case references
+        </label>
+    </p>
+    <p>
+        "Breaking @OrgConfig.get().str_res_plan" entry:
```
[the remark this answers is not in the document]
Comment (aaronbrowder) [reply app/views/view_settings.scala.html]: It sounds like it's failing to save your advanced JC settings. Can you confirm that in the db?

File: app/controllers/Application.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaronbrowder) [reply app/controllers/Application.java]: No, it's just pulling display_name directly from the db. I will change it to use getDisplayName() and maybe that will fix it.

[the remark this answers is not in the document]
Comment (aaronbrowder) [reply app/controllers/Application.java]: So the `display_name` issue is not causing the exception. That is a NullPointerException on the line:
`return Application.formatDayOfWeek(the_case.meeting.date);`
in the context of the newly generated charge. So either `the_case` is null or `the_case.meeting` is null on the newly generated charge. Could you send a snapshot of the db row of the generated charge so I can make sure it generated correctly?

**Comment by aaronbrowder** [comment]
When you reference a case, you see the original case details above the findings field. This allows you to write findings based on previous charges in the event that a sentence was broken.

image

When you click "Generate charge" it creates a charge with the person prefilled. The rule is also prefilled if the "breaking sentence entry" setting is set. Prefilled items cannot be deleted.

image

For the sentence, you have the option of choosing "time served" or creating a new sentence. Either way, the original sentence is automatically marked completed.

image

Anywhere the findings of a case are rendered, it renders the full sequence of events.

image

If a referenced case has multiple charges, but only some of those charges are relevant in the new case, you can uncheck the irrelevant ones and they will be excluded from the report.

image

**Comment by aaronbrowder** [comment]
I made the requested changes, and added a null check which should prevent the exception you got. I'm still not sure why the case or meeting was null, so there may be other issues, but possibly not.

**Comment by schmave** [comment]
Awesome, thanks! I discovered that my local build was failing in a variety of weird ways that were fixed once I ran an "sbt clean", so I expect that the issues will be better now. I will take another look very soon.

**Comment by schmave** [comment]
Hi Aaron,

I took another look today and I have a couple more questions. 

Would it be OK to replace the checkbox option label "Time served" with "No new sentence"? I'm not sure I'm understanding "Time served" right, but if it is equivalent to "no new sentence", then the system shouldn't generate a new resolution plan with the text "Time served". What do you think?

Would it be OK to replace the label "Original:" with "Previous sentence:"? I did this in code already, but am happy to change it to something else.

I think it's a nice feature that the existing sentence is marked as complete when a referenced charge is created, but I have a couple issues with it. One is that some school may not treat broken sentences that way and there's no way to choose. The other is that if you delete the generated charge, the original charge's sentence is not unmarked complete. A simple way to fix both issues is to not do the auto-complete when a new charge is generated, but maybe you have another idea.

**Comment by aaronbrowder** [comment]
There's a good reason I decided to mark the original sentence as complete and use the term "time served". My school actually does not treat broken sentences this way; in fact we have not had consistency in the way we treat broken sentences, and that's one of the problems this project is supposed to solve.

Imagine that Sam has a sentence to do a chore. He completes the chore halfway and gives up, and is then written up. JC charges him with breaking sentence and assigns the sentence "1 additional chore". Now as a person reading this report, does Sam now have to do 2 chores or only 1 chore? JC did not specify whether the original chore was counted as completed, although that may have been clear to the members of JC while discussing it.

Now imagine a more difficult scenario: Sam has a sentence that he is restricted from the playground for 2 days. He completes his first day, and then on the second day in the morning he plays on the playground and is written up. JC charges him with breaking sentence and assigns the sentence "1 additional day restricted from playground". Now as a person reading this report, is Sam only restricted from the playground tomorrow, or tomorrow and the next day? Perhaps JC intended that it would be tomorrow and the next day since he technically didn't complete his original 2nd day, but the sentence "1 additional day restricted from playground" does not convey that and can possibly convey the opposite.

The new feature attempts to remove the ambiguity by removing human agency from the sentencing process. The user only has the option of "time served" (which clearly conveys that the defendant is off the hook and doesn't have to do anything more) vs. "replace sentence" (which clearly conveys that the original sentence is null and must be replaced). There is no option to retain the original sentence and modify or supplement it, which would open the door for ambiguity.

> I'm not sure I'm understanding "Time served" right, but if it is equivalent to "no new sentence", then the system shouldn't generate a new resolution plan with the text "Time served".

It's not possible to create a charge with no resolution plan, since they occupy a single entity. It is possible to automatically mark "time served" sentences as completed, and I would not object to the system doing that.

> Would it be OK to replace the label "Original:" with "Previous sentence:"?

I used the label "Original" so it wouldn't wrap. "Previous sentence" is fine as long as it looks okay to you.

**Comment by schmave** [comment]
Thanks for the extra explanation. Those examples are very helpful. I like the idea about being explicit about whether the existing/broken sentence is still in effect or not. I think I would like that to be even clearer in the interface. Maybe you/I could autofill the text of the sentence to say that the previous sentence is no longer in effect. It's a little too hidden right now for my taste.

The bug I'm talking about is still an issue, though -- if you click "generate charge" and then change your mind and click "delete charge" on the generated charge, the original RP is still marked as complete.

**Comment by aaronbrowder** [comment]
Sure, you can add that prefilled text if you're already working on the code. I agree that it's too hidden.

How would you propose handling the bug? It could be tricky reversing the completion, especially if the charge is deleted a while after it's created. If they actually want the sentence to be marked completed, then silently un-completing it would be problematic. What if it just stores the previous value in javascript when a charge is created and then reverts it if they delete the charge in the same page load?

**Comment by aaronbrowder** [comment]
Actually I have a better idea: instead of marking the sentence complete, we can have it be implicitly complete by the fact that it is referenced. So anywhere we are reading `rp_complete`, we can instead read `rp_complete || is_referenced`.

**Comment by schmave** [comment]
Genius!

**Comment by aaronbrowder** [comment]
OK, I made the change I suggested. Now if a charge is referenced, its sentence is considered "nullified" and appears in a 3rd section on the sentences page titled "recently nullified sentences".

I also added a note by "previous sentence" which says "_Note: This charge nullifies the previous sentence. You need to replace it or choose "time served"_."

**Comment by schmave** [comment]
Thanks for making those changes, @aaronbrowder. I made a few changes, mostly to wording and layout. It's been deployed now. Give it a try and let me know how it goes. I'll send an email to the users list in a week or so about the new features. 

Thanks again for taking the time to get these features merged!

**Comment by aaronbrowder** [comment]
Thanks for your help! We'll try out the feature over the next few days.

**Comment by aaronbrowder** [comment]
So far we haven't had many opportunities to test this and find issues, so you may want to wait a bit before announcing it to everyone.

I noticed that you changed the string "time served" to "no sentence" or "none". Is that because "time served" is inappropriate for schools that don't use the terminology "sentence"? If so, would it be possible to get that changed to "time served" just for The Open School? We feel that "no sentence" and "none" do not accurately convey that the original sentence has been erased, and may be interpreted as "no additional sentence".

**Comment by schmave** [comment]
Yeah, I didn't like the phrase "time served". But if you want "time served", you should have it. I'll make it configurable later if another school asks for it.

**Comment by aaronbrowder** [comment]
Thank you

**Pull request merged**